### PR TITLE
[FLINK-20702][python] Support Map Operation Chained Together in Python Table API

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will Support Map Operation Chained Together in Python Table API*

## Brief change log

  - *Add logic rewrite rule `PythonCalcMergeRule`*

## Verifying this change

This change added tests and can be verified as follows:

  - *UT `PythonCalcMergeRuleTest`*
  - *IT `test_map_with_pandas_udf`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/table/tests/test_row_based_operation.py [hunk 1]
```diff
@@ -67,17 +67,28 @@ def test_map_with_pandas_udf(self):
 
         def func(x):
             import pandas as pd
-            res = pd.concat([x.a, x.c + x.d], axis=1)
+            b = (x.c + x.d).rename('b')
```
Comment (WeiZhong94) [inline flink-python/pyflink/table/tests/test_row_based_operation.py:70]: unnecessary change.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+			return false;
+		}
+
+		RexProgram middleProgram = middleCalc.getProgram();
+		List<RexNode> middleProjects = middleProgram.getProjectList()
+			.stream()
+			.map(middleProgram::expandLocalRef)
+			.collect(Collectors.toList());
+
+		return isInputsCorrespondWithUpstreamOutputs(bottomProjects, middleProjects) &&
```
Comment (WeiZhong94) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:89]: We should ensure the `condition` field of the `middleProgram ` is null here.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		}
+		return true;
+	}
+
+	private boolean isFlattenCalc(List<RexNode> middleProjects) {
+		for (RexNode middleProject : middleProjects) {
+			if (middleProject instanceof RexFieldAccess) {
+				RexNode expr = ((RexFieldAccess) middleProject).getReferenceExpr();
+				if (expr instanceof RexInputRef) {
+					if (((RexInputRef) expr).getIndex() != 0) {
```
Comment (WeiZhong94) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:117]: This condition is too loose to determine whether the calc is a flatten calc.

**Review by WeiZhong94 — COMMENTED** [review]
@HuangXingBo Thanks for your PR! There are some comments for you.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java (hunk 2 above)
Comment (HuangXingBo) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:89]: Yes. Good catch.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java (hunk 3 above)
Comment (HuangXingBo) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:117]: Yes. Good Catch.

**Comment by HuangXingBo** [comment]
@WeiZhong94 Thanks a lot for the review. I have addressed the comments at the latest commit.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Rule will merge Python {@link FlinkLogicalCalc} used in Map operation, Flatten {@link FlinkLogicalCalc}
+ * and Python {@link FlinkLogicalCalc} used in Map operation together.
+ */
+public class PythonCalcMergeRule extends RelOptRule {
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:43]: The name may cause confusing as CalcMerge usually means merging two Calc into one Calc. Besides, the name doesn't reflect the real purpose: merging two continuous map operation

If so, what about rename it to PythonMapMergeRule or some other name?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+			.getFieldList()
+			.get(0)
+			.getValue()
+			.getFieldList().size();
+
+		return isInputsCorrespondWithUpstreamOutputs(bottomProjects, middleProjects) &&
+			isFlattenCalc(middleProjects, inputRowFieldCount);
+	}
+
+	private boolean isInputsCorrespondWithUpstreamOutputs(List<RexNode> bottomProjects, List<RexNode> middleProjects) {
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:102]: I guess we should also check whether the topCalc takes row as input.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+		FlinkLogicalCalc middleCalc = call.rel(1);
+		FlinkLogicalCalc topCalc = call.rel(2);
+
+		RexProgram bottomProgram = bottomCalc.getProgram();
+		List<RexCall> bottomProjects = bottomProgram.getProjectList()
+			.stream()
+			.map(bottomProgram::expandLocalRef)
+			.map(x -> (RexCall) x)
+			.collect(Collectors.toList());
+		RexCall bottomPythonCall = bottomProjects.get(0);
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:159]: Should we also check whether there is condition in the bottomCalc and topCalc?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+	private PythonCalcMergeRule() {
+		super(operand(FlinkLogicalCalc.class,
+			operand(FlinkLogicalCalc.class,
+				operand(FlinkLogicalCalc.class, none()))),
+			"PythonCalcMergeRule");
+	}
+
+	@Override
+	public boolean matches(RelOptRuleCall call) {
+		FlinkLogicalCalc bottomCalc = call.rel(0);
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:56]: I will change the names

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java (hunk 4 above)
Comment (HuangXingBo) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:43]: Yes. `PythonMapMergeRule` would be better

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java (hunk 6 above)
Comment (HuangXingBo) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java:159]: `bottomCalc` and `topCalc` are PythonCalc which won't contain condition. so we don't need to check condition.

**Review by dianfu — APPROVED** [review]
LGTM.
